sagemath/sage
# Preliminary steps to save the CI infrastructure

To improve the situation with the CI infrastructure, this PR:

- added comments untangling obscure code in CI-related files, for those poor guys who ever attempt to read the files for whatever reasons.

- while doing the cosmetic changes, a bug (about `-uninstall` targets) was found `build/make/Makefile.in`, which is fixed here.

  to test, do
  ```
  $ ./configure --enable-dot2tex | grep dot2tex
  $ make build | grep dot2tex
  $ ./configure --disable-dot2tex | grep dot2tex
  $ make build | grep dot2tex
  ```

- fixed some jobs in the CI-linux workflow that fail because of duplicate artifact names.

- removed ubuntu-lunar, ubuntu-mantic, conda-forge-python3.11, ubuntu-bionic-gcc_8-i386, debian-bullseye-i386 from the list of the default systems that CI runs for. This is how to properly modify the list:
   - first edit `tox.ini` (find DEFAULT_SYSTEM_FACTORS)
   - run `tox -e update_docker_platforms`
   - commit the changes

- removed old versions of linuxmint and added new versions.

- removed old versions of fedora distributions.

- "optional" and "experimental" jobs now run upon "standard" docker images, instead of "maximal" ones, to avoid "out of runner space" error.

- renamed "Reusable workflow for Docker-based portability CI" to "Workflow for Linux portability CI" for short name and made it runnable through github interface to facilitate testing specific platform by adding "workflow-dispatch" calling `docker.yml`.

   test: https://github.com/kwankyu/sage/actions/workflows/docker.yml

- added helpful comments and updated the developer doc

- reimplemented `.ci/write-dockerfile.sh` so that simplified Dockerfile is generated for present and future stability

- turned off failing jobs in "CI Linux incremental"

- removed seemingly useless `subprojects/factory` directory to eliminate certain git warnings. 

- turned off "standard-sitepackegs" and "standard-constraints_pkgs-norequirements" jobs as they fail on (almost) all platforms.

test CI run (as of 10.6.beta8): https://github.com/kwankyu/sage/actions/runs/13692352790
compare with the status quo: https://github.com/sagemath/sage/actions/runs/13596179432

test CI with a PR: https://github.com/kwankyu/sage/pull/82

The main objective of this PR is to solve issues with the workflow "CI Linux" such that a failure on a platform reveals solely some problem of sage built on the platform, but not a problem of the CI infrastructure. After this PR, hopefully, each of failing platforms should be tackled individually. If a platform fails, perhaps we should

1. decide first whether to support the platform or not. 
2. if the platform is supported, open a github issue for it.
3. if the platform is not supported, then remove it from the "master list of supported linux platforms" in tox.ini.
4. if a supported platform constantly fails but no PR for the issue is present, then we may turn it off (by commenting it out) until fixed.

I suggest discontinuing support (at least in CI) for Linux releases that have been past their EOL (end of life or end of support by the distributor) for more than 2 years.

Only decent platforms according to the CI results should be listed in https://github.com/sagemath/sage/wiki/Sage-10.6-Release-Tour#availability-and-installation-help.

The following diagram shows how packages are installed for each of CI jobs:
```
                _prereq | standard package | optional package | experimental package
-------------------------------------------------------------------------------------
"minimal"        SSSSSS | ---------------- |
"standard"       SSSSSS | SSSSSSSSSSS----- |
"maximal"        SSSSSS | SSSSSSSSSSS----- | SSSS------------ | 
"optional"       SSSSSS | SSSSSSSSSSS----- | ---------------- | 
"experimental"   SSSSSS | SSSSSSSSSSS----- |                  | ------------------ |
```
where "S" represents system package and dash "-" represents Sage package. Hence

- In the test results of  "minimal" job, we can examine how well standard **sage packages** behave with sage.
- In the test results of  "standard" job, we can examine how well standard **system packages** behave with sage.
- In the test results of  "maximal" job, we can examine how well optional **system packages** behave with sage.
- In the test results of  "optional" job, we can examine how well optional **sage packages** behave with sage. 
- In the test results of  "experimental" job, we can examine how well experimental **sage packages** behave with sage. 

### :memo: Checklist



- [x] The title is concise and informative.
- [x] I have updated the documentation and checked the documentation preview.

### :hourglass: Dependencies
